### ISSUE TITLE
Analyzer: warns only about using const constructor when const is used at least once

### ISSUE THREAD

**Issue by goderbauer** [dart-lang/linter#1712]
```
$ dartanalyzer --version
dartanalyzer version 2.6.0-edge.36985859e421a9fbbdfc56379506b2a367c7d4e9
```

I have the `prefer_const_constructors` lint enabled. Running the analyzer over this file:

```dart
void main() {
  Duration(seconds: 10);
  Duration(seconds: 10);
  Duration(seconds: 10);
  Duration(seconds: 10);
  Duration(seconds: 10);
  Duration(seconds: 10);
  Duration(seconds: 10);
}
```

shows no issues:

```
$ dartanalyzer main.dart 
Analyzing main.dart...
No issues found!
```

That is already strange as all the Durations should be const! Adding const to one of the Durations suddenly makes the analyzer detect that all the Durations should be const as well:

```dart
void main() {
  const Duration(seconds: 10); // Note the const here
  Duration(seconds: 10);
  Duration(seconds: 10);
  Duration(seconds: 10);
  Duration(seconds: 10);
  Duration(seconds: 10);
  Duration(seconds: 10);
}
```

```
$ dartanalyzer main.dart 
Analyzing main.dart...
  lint • Prefer const with constant constructors. • main.dart:3:3 • prefer_const_constructors
  lint • Prefer const with constant constructors. • main.dart:4:3 • prefer_const_constructors
  lint • Prefer const with constant constructors. • main.dart:5:3 • prefer_const_constructors
  lint • Prefer const with constant constructors. • main.dart:6:3 • prefer_const_constructors
  lint • Prefer const with constant constructors. • main.dart:7:3 • prefer_const_constructors
  lint • Prefer const with constant constructors. • main.dart:8:3 • prefer_const_constructors
6 lints found.
```

**Comment by pq** [comment]
Thanks for the report!

@scheglov: I wonder if you have any insights into this one?   (We're deferring to `LinterContextImpl.canBeConst`):

https://github.com/dart-lang/linter/blob/f1bce463d96b1af6b02da1fe220ddbce6ef793a3/lib/src/rules/prefer_const_constructors.dart#L95-L103

(We should also remove the extra visitor if it's no longer needed...)

**Comment by scheglov** [comment]
https://dart-review.googlesource.com/c/sdk/+/115613 for `Duration`.

**Comment by scheglov** [comment]
As for type parameters, the reason is that `CompileTimeErrorCode.CONST_WITH_TYPE_PARAMETERS` is reported in `ErrorVerifier`, not during constant evaluation.

**Comment by pq** [comment]
@scheglov: is this fixed as of [2246f0a](https://dart.googlesource.com/sdk/+/2246f0a8a36fbad0cd9c87aacc22ba29086d1c74)?

**Comment by scheglov** [comment]
Yes, the title issue is fixed.

I did not close it yet because I wanted to get back to not reporting errors for "constants" with type parameters.

**Comment by pq** [comment]
Ok cool.  Thank you!

**Comment by scheglov** [comment]
With c11ca7d4395d5ca2f2df3349e7ab92b320021fce we can remove the `_Collector` work-around.
Obviously this means that you need a new analyzer ;-)

**Comment by pq** [comment]
Fantastic!  @scheglov: do we need a new `analyzer` published w/ the required fixes?

**Comment by scheglov** [comment]
@pq Yes, we do need a new version of `analyzer`, for this fix, to remove `_Collector`, and hopefully also to remove the `MockDartSdk` from the linter. But I'm also making to changes to `TypeProvider` and the way `ClassElement`(s) are turned into `InterfaceType`(s). So, we might publish `analyzer` and switch `linter` to it, but we will need another publishing and `linter` changes soonish.
